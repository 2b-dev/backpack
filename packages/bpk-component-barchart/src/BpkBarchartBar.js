import React from 'react';
import PropTypes from 'prop-types';
import { borderRadiusXs } from 'bpk-tokens/tokens/base.es6';
import { remToPx } from './utils';

const KEYCODES = {
  ENTER: 13,
  SPACEBAR: 32,
};

const handleKeyboardEvent = callback => (event) => {
  if (event.keyCode === KEYCODES.ENTER || event.keyCode === KEYCODES.SPACEBAR) {
    event.preventDefault();
    callback(event);
  }
};

const borderRadius = remToPx(borderRadiusXs);

const BpkBarchartBar = (props) => {
  const {
    x,
    y,
    width,
    height,
    className,
    label,
    onClick,
    outlier,
    ...rest
  } = props;

  const classNames = ['bpk-barchart__bar'];
  if (className) { classNames.push(className); }
  if (outlier) { classNames.push('bpk-barchart__bar--outlier'); }
  if (onClick) { classNames.push('bpk-barchart__bar--interactive'); }

  return (
    <g className="bpk-barchart__bar-group">
      <title>{label}</title>
      <rect
        className={classNames.join(' ')}
        x={x}
        y={y}
        width={width}
        height={height}
        rx={borderRadius}
        ry={borderRadius}
        onClick={onClick || undefined}
        onKeyDown={onClick ? handleKeyboardEvent(onClick) : undefined}
        tabIndex={onClick ? 0 : undefined}
        role={onClick ? 'button' : undefined}
        aria-label={label}
        {...rest}
      />
    </g>
  );
};

BpkBarchartBar.propTypes = {
  height: PropTypes.number.isRequired,
  label: PropTypes.string.isRequired,
  width: PropTypes.number.isRequired,
  x: PropTypes.number.isRequired,
  y: PropTypes.number.isRequired,
  className: PropTypes.string,
  onClick: PropTypes.func,
  outlier: PropTypes.bool,
};

BpkBarchartBar.defaultProps = {
  className: null,
  onClick: null,
  outlier: false,
};

export default BpkBarchartBar;
