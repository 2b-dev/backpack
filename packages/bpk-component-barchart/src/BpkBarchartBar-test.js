import React from 'react';
import renderer from 'react-test-renderer';
import BpkBarchartBar from './BpkBarchartBar';

describe('BpkBarchartBar', () => {
  it('should render correctly', () => {
    const tree = renderer.create(
      <BpkBarchartBar
        x={10}
        y={10}
        width={20}
        height={100}
        label="Bar"
      />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render with an onClick handler', () => {
    const tree = renderer.create(
      <BpkBarchartBar
        x={10}
        y={10}
        width={20}
        height={100}
        label="Bar"
        onClick={() => null}
      />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render as an outlier', () => {
    const tree = renderer.create(
      <BpkBarchartBar
        x={10}
        y={10}
        width={20}
        height={100}
        label="Bar"
        outlier
      />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
