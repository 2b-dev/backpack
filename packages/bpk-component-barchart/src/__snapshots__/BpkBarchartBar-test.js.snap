// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BpkBarchartBar should render as an outlier 1`] = `
<g
  className="bpk-barchart__bar-group"
>
  <title>
    Bar
  </title>
  <rect
    aria-label="Bar"
    className="bpk-barchart__bar bpk-barchart__bar--outlier"
    height={100}
    onClick={undefined}
    onKeyDown={undefined}
    role={undefined}
    rx={3}
    ry={3}
    tabIndex={undefined}
    width={20}
    x={10}
    y={10}
  />
</g>
`;

exports[`BpkBarchartBar should render correctly 1`] = `
<g
  className="bpk-barchart__bar-group"
>
  <title>
    Bar
  </title>
  <rect
    aria-label="Bar"
    className="bpk-barchart__bar"
    height={100}
    onClick={undefined}
    onKeyDown={undefined}
    role={undefined}
    rx={3}
    ry={3}
    tabIndex={undefined}
    width={20}
    x={10}
    y={10}
  />
</g>
`;

exports[`BpkBarchartBar should render with an onClick handler 1`] = `
<g
  className="bpk-barchart__bar-group"
>
  <title>
    Bar
  </title>
  <rect
    aria-label="Bar"
    className="bpk-barchart__bar bpk-barchart__bar--interactive"
    height={100}
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    rx={3}
    ry={3}
    tabIndex={0}
    width={20}
    x={10}
    y={10}
  />
</g>
`;
